src/data/complex/R_or_C.lean

```diff
@@ -0,0 +1,613 @@
+  λ z w hre him, is_R_or_C.ext_iff.mpr ⟨hre, him⟩
+
+
+lemma zero_re : re (𝓚 0) = (0 : ℝ) := by simp only [of_real_re]
+@[simp] lemma zero_im : im (𝓚 0) = 0 := by rw [of_real_im]
+lemma of_real_zero : 𝓚 0 = 0 := by rw [of_real_alg, zero_smul]
+
+@[simp] lemma of_real_one : 𝓚 1 = 1 := by rw [of_real_alg, one_smul]
+@[simp] lemma one_re : re (1 : K) = 1 := by rw [←of_real_one, of_real_re]
+@[simp] lemma one_im : im (1 : K) = 0 := by rw [←of_real_one, of_real_im]
+
+@[simp] theorem of_real_inj {z w : ℝ} : 𝓚 z = 𝓚 w ↔ z = w :=
+begin
+  split,
+  { intro h,
+    have := congr_arg re h,
+    simp only [of_real_re] at this,
+    exact this },
+  { intro h,
+    simp only [h] }
+end
+
+@[simp] lemma bit0_re (z : K) : re (bit0 z) = bit0 (re z) := by simp [bit0]
+@[simp] lemma bit1_re (z : K) : re (bit1 z) = bit1 (re z) :=
+  by simp only [bit1, add_monoid_hom.map_add, bit0_re, add_right_inj, one_re]
+@[simp] lemma bit0_im (z : K) : im (bit0 z) = bit0 (im z) := by simp [bit0]
+@[simp] lemma bit1_im (z : K) : im (bit1 z) = bit0 (im z) :=
+  by simp only [bit1, add_right_eq_self, add_monoid_hom.map_add, bit0_im, one_im]
+
+@[simp] theorem of_real_eq_zero {z : ℝ} : 𝓚 z = 0 ↔ z = 0 :=
+  by rw [←of_real_zero]; exact of_real_inj
+
+@[simp] lemma of_real_add (r s : ℝ) : 𝓚 (r + s) = 𝓚 r + 𝓚 s :=
+  by apply (@is_R_or_C.ext_iff K _ _ _ (𝓚 (r + s)) (𝓚 r + 𝓚 s)).mpr; simp
+
+@[simp] lemma of_real_bit0 (r : ℝ) : 𝓚 (bit0 r : ℝ) = bit0 (𝓚 r) :=
+ext_iff.2 $ by simp [bit0]
+
+@[simp] lemma of_real_bit1 (r : ℝ) : 𝓚 (bit1 r : ℝ) = bit1 (𝓚 r) :=
+ext_iff.2 $ by simp [bit1]
+
+lemma two_eq_of_real : (2 : K) = 𝓚 2 := by rw [bit0, ←of_real_one, ←of_real_add, bit0]
+
+lemma two_ne_zero : (2 : K) ≠ 0 :=
+begin
+  intro h,
+  apply (show (2 : ℝ) ≠ 0, by linarith),
+  rw [two_eq_of_real, ←of_real_zero] at h,
+  exact of_real_inj.mp h,
+end
+
+@[simp] lemma of_real_neg (r : ℝ) : 𝓚 (-r : ℝ) = -(𝓚 r) := ext_iff.2 $ by simp
+@[simp] lemma of_real_mul (r s : ℝ) : 𝓚 (r * s : ℝ) = (𝓚 r) * (𝓚 s) := ext_iff.2 $ by simp
+
+lemma smul_re (r : ℝ) (z : K) : re ((𝓚 r) * z) = r * (re z) :=
+  by simp only [of_real_im, zero_mul, of_real_re, sub_zero, mul_re]
+lemma smul_im (r : ℝ) (z : K) : im ((𝓚 r) * z) = r * (im z) :=
+  by simp only [add_zero, of_real_im, zero_mul, of_real_re, mul_im]
+
+lemma smul_re' : ∀ (r : ℝ) (z : K), re (r • z) = r * (re z) :=
+  λ r z, by rw [smul_coe_mul_ax]; exact smul_re r z
+lemma smul_im' : ∀ (r : ℝ) (z : K), im (r • z) = r * (im z) :=
+  λ r z, by rw [smul_coe_mul_ax]; exact smul_im r z
+
+/-! ### The imaginary unit, `I` -/
+
+/-- The imaginary unit. -/
+@[simp] lemma I_re : re (I : K) = 0 := I_re_ax
+lemma I_mul_I : (I : K) = 0 ∨ (I : K) * I = -1 := I_mul_I_ax
+
+@[simp] lemma conj_re (z : K) : re (conj z) = re z := is_R_or_C.conj_re_ax z
+@[simp] lemma conj_im (z : K) : im (conj z) = -(im z) := is_R_or_C.conj_im_ax z
+@[simp] lemma conj_of_real (r : ℝ) : conj (𝓚 r) = (𝓚 r) :=
+  (@is_R_or_C.ext_iff K _ _ _ (conj (𝓚 r)) (𝓚 r)).mpr $ by simp
```

Comment:
```suggestion
by { rw ext_iff, simp only [of_real_im, conj_im, eq_self_iff_true, conj_re, and_self, neg_zero] }
```
squeezing this simp makes the file compile a bit faster. I don't have a profiler output, but the orange bar hangs around noticeably long with a raw `simp` and goes away immediately with this `simp only`

Reply:
Actually the proofs of conj_neg_I and conj_conj take a while, too. Maybe we've made a bad choice with our simp normal form?

Reply:
I haven't been too careful with the simp lemmas, I basically left things the way they were in `complex.lean`.